src/MSALWrapper/AuthFlow/Broker.cs

```diff
@@ -73,65 +70,46 @@ private enum GetAncestorFlags
+        protected override async Task<TokenResult> GetTokenInnerAsync()
         {
             IAccount account = await this.pcaWrapper.TryToGetCachedAccountAsync(this.preferredDomain)
                  ?? PublicClientApplication.OperatingSystemAccount;
 
-            TokenResult tokenResult = null;
-            try
-            {
-                tokenResult = await CachedAuth.GetTokenAsync(
-                    this.logger,
-                    this.scopes,
-                    account,
-                    this.pcaWrapper,
-                    this.errors);
-
-                if (tokenResult != null)
-                {
-                    return (tokenResult, this.errors);
-                }
+            TokenResult tokenResult = await CachedAuth.GetTokenAsync(
+                this.logger,
+                this.scopes,
+                account,
+                this.pcaWrapper,
+                this.errors);
 
-                try
-                {
-                    tokenResult = await TaskExecutor.CompleteWithin(
-                        this.logger,
-                        this.interactiveAuthTimeout,
-                        $"{this.Name()} interactive auth",
-                        this.GetTokenInteractive(account),
-                        this.errors).ConfigureAwait(false);
-                }
-                catch (MsalUiRequiredException ex)
-                {
-                    this.errors.Add(ex);
-                    this.logger.LogDebug($"Initial {this.Name()} auth failed. Trying again with claims from exception.\n{ex.Message}");
-
-                    tokenResult = await TaskExecutor.CompleteWithin(
-                        this.logger,
-                        this.interactiveAuthTimeout,
-                        $"{this.Name()} interactive auth (with extra claims)",
-                        this.GetTokenInteractiveWithClaims(ex.Claims),
-                        this.errors).ConfigureAwait(false);
-                }
-            }
-            catch (MsalServiceException ex)
+            if (tokenResult != null)
             {
-                this.logger.LogWarning($"MSAL Service Exception! (Not expected)\n{ex.Message}");
-                this.errors.Add(ex);
+                return tokenResult;
             }
-            catch (MsalClientException ex)
+
+            try
             {
-                this.logger.LogWarning($"Msal Client Exception! (Not expected)\n{ex.Message}");
-                this.errors.Add(ex);
+                tokenResult = await TaskExecutor.CompleteWithin(
+                    this.logger,
+                    this.interactiveAuthTimeout,
+                    $"{this.Name()} interactive auth",
+                    this.GetTokenInteractive(account),
+                    this.errors).ConfigureAwait(false);
             }
-            catch (NullReferenceException ex)
+            catch (MsalUiRequiredException ex)

```

Comment:
Line 101, do we still need to add exceptions to the local errors list? I'm trying to figure out how these errors are propagated to the base class. I see that any `MsalUiRequiredException` from here wouldn't propagate up. Am I missing something?

Reply:
Previously we had two levels of `catch (MsalUIRequiredException)` directly nested. We still have that same concept, but now the outer catch happens in the base class method `GetTokenAsync`. This inner catch is still required for `broker` and `web` modes where we might need to try doing interactive auth twice, the second time with extra claims returned in the exception. When we call again with those extra claims, an MSAL UI Required exception can be thrown again, if say, the error was not about claims, but instead that MFA was required, and they failed to pass it.

We could probably streamline this better in the future by checking if there are claims in the exception - and only do this extra attempt if there are claims. If not, we could rethrow and let the base class handle catching and logging it.

Reply:
@mvanchaa I'm going to merge to keep going on these, but circle back please if this doesn't answer your question and we can discuss :)

Reply:
Thanks for explaining @kyle-rader-msft. You answered my question.